server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentTracker.java

```diff
@@ -0,0 +1,425 @@
+     */
+    private final AtomicReference<MovingAverage> uploadBytesPerSecMovingAverageReference;
+
+    private final Object uploadBytesPerSecMutex = new Object();
+
+    /**
+     * Provides moving average over the last N overall upload time (in nanos) as part of remote refresh.N is window size.
+     * Wrapped with {@code AtomicReference} for dynamic changes in window size.
+     */
+    private final AtomicReference<MovingAverage> uploadTimeMsMovingAverageReference;
+
+    private final Object uploadTimeMsMutex = new Object();
+
+    public RemoteRefreshSegmentTracker(
+        ShardId shardId,
+        int uploadBytesMovingAverageWindowSize,
+        int uploadBytesPerSecMovingAverageWindowSize,
+        int uploadTimeMsMovingAverageWindowSize
+    ) {
+        this.shardId = shardId;
+        // Both the local refresh time and remote refresh time are set with current time to give consistent view of time lag when it arises.
+        long currentTimeMs = System.nanoTime() / 1_000_000L;
+        localRefreshTimeMs = currentTimeMs;
+        remoteRefreshTimeMs = currentTimeMs;
+        uploadBytesMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadBytesMovingAverageWindowSize));
+        uploadBytesPerSecMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadBytesPerSecMovingAverageWindowSize));
+        uploadTimeMsMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadTimeMsMovingAverageWindowSize));
+    }
+
+    ShardId getShardId() {
+        return shardId;
+    }
+
+    long getLocalRefreshSeqNo() {
+        return localRefreshSeqNo;
+    }
+
+    void updateLocalRefreshSeqNo(long localRefreshSeqNo) {
+        assert localRefreshSeqNo > this.localRefreshSeqNo : "newLocalRefreshSeqNo="
+            + localRefreshSeqNo
+            + ">="
+            + "currentLocalRefreshSeqNo="
+            + this.localRefreshSeqNo;
+        this.localRefreshSeqNo = localRefreshSeqNo;
+        computeSeqNoLag();
+    }
+
+    long getLocalRefreshTimeMs() {
+        return localRefreshTimeMs;
+    }
+
+    void updateLocalRefreshTimeMs(long localRefreshTimeMs) {
+        assert localRefreshTimeMs > this.localRefreshTimeMs : "newLocalRefreshTimeMs="
+            + localRefreshTimeMs
+            + ">="
+            + "currentLocalRefreshTimeMs="
+            + this.localRefreshTimeMs;
+        this.localRefreshTimeMs = localRefreshTimeMs;
+        computeTimeMsLag();
+    }
+
+    long getRemoteRefreshSeqNo() {
+        return remoteRefreshSeqNo;
+    }
+
+    void updateRemoteRefreshSeqNo(long remoteRefreshSeqNo) {
+        assert remoteRefreshSeqNo > this.remoteRefreshSeqNo : "newRemoteRefreshSeqNo="
+            + remoteRefreshSeqNo
+            + ">="
+            + "currentRemoteRefreshSeqNo="
+            + this.remoteRefreshSeqNo;
+        this.remoteRefreshSeqNo = remoteRefreshSeqNo;
+        computeSeqNoLag();
```

Comment:
lets call it `RefreshSeqNoLag` here and also at other places.

Reply:
addressed